File: frontend/src/types/comparison.ts
export type ComparisonNamedEntity = {
    id: string;
    name: string;
};

export type ComparisonExhibition =
    ComparisonNamedEntity & {
    year: number | null;
};

export type ComparisonArtist = {
    id: string;
    entity: string;
    display_name: string;

    cluster: number;
    is_noise: boolean;
    membership_probability: number;
    outlier_score: number;

    birth_year: number | null;
    death_year: number | null;
    nationality: string | null;
    gender: string | null;

    groups: ComparisonNamedEntity[];
    exhibitions: ComparisonExhibition[];
    locations: ComparisonNamedEntity[];
};

export type ComparisonDifference = {
    label: string;
    artist_a_value: string;
    artist_b_value: string;
    same: boolean;
};

export type ComparisonPathNode = {
    key: string;
    id: string;
    label: string;
    type: string;
    properties: Record<string, unknown>;
};

export type ComparisonPathLink = {
    source: string;
    target: string;
    type: string;
};

export type ComparisonShortestPath = {
    found: boolean;
    hops: number | null;
    nodes: ComparisonPathNode[];
    links: ComparisonPathLink[];
};

export type ArtistComparisonResponse = {
    artist_a: ComparisonArtist;
    artist_b: ComparisonArtist;

    embedding_similarity: number;
    same_cluster: boolean;

    common: {
        groups: ComparisonNamedEntity[];
        exhibitions: ComparisonExhibition[];
        locations: ComparisonNamedEntity[];
    };

    unique: {
        artist_a_groups: ComparisonNamedEntity[];
        artist_b_groups: ComparisonNamedEntity[];
        artist_a_exhibitions: ComparisonExhibition[];
        artist_b_exhibitions: ComparisonExhibition[];
        artist_a_locations: ComparisonNamedEntity[];
        artist_b_locations: ComparisonNamedEntity[];
    };

    differences: ComparisonDifference[];

    co_exhibited: {
        direct_relationship: boolean;
        shared_exhibition_count: number;
    };

    shortest_path: ComparisonShortestPath;
    note: string;
};

File: backend/app/schemas/comparison.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class ComparisonNamedEntity(BaseModel):
    id: str
    name: str


class ComparisonExhibition(ComparisonNamedEntity):
    year: int | None = None


class ComparisonArtist(BaseModel):
    id: str
    entity: str
    display_name: str
    cluster: int
    is_noise: bool
    membership_probability: float
    outlier_score: float
    birth_year: int | None = None
    death_year: int | None = None
    nationality: str | None = None
    gender: str | None = None
    groups: list[ComparisonNamedEntity] = Field(default_factory=list)
    exhibitions: list[ComparisonExhibition] = Field(default_factory=list)
    locations: list[ComparisonNamedEntity] = Field(default_factory=list)


class ComparisonCommon(BaseModel):
    groups: list[ComparisonNamedEntity] = Field(default_factory=list)
    exhibitions: list[ComparisonExhibition] = Field(default_factory=list)
    locations: list[ComparisonNamedEntity] = Field(default_factory=list)


class ComparisonUnique(BaseModel):
    artist_a_groups: list[ComparisonNamedEntity] = Field(default_factory=list)
    artist_b_groups: list[ComparisonNamedEntity] = Field(default_factory=list)
    artist_a_exhibitions: list[ComparisonExhibition] = Field(default_factory=list)
    artist_b_exhibitions: list[ComparisonExhibition] = Field(default_factory=list)
    artist_a_locations: list[ComparisonNamedEntity] = Field(default_factory=list)
    artist_b_locations: list[ComparisonNamedEntity] = Field(default_factory=list)


class ComparisonDifference(BaseModel):
    label: str
    artist_a_value: str
    artist_b_value: str
    same: bool


class CoExhibitedSummary(BaseModel):
    direct_relationship: bool
    shared_exhibition_count: int


class ComparisonPathNode(BaseModel):
    key: str
    id: str
    label: str
    type: str
    properties: dict[str, Any] = Field(default_factory=dict)


class ComparisonPathLink(BaseModel):
    source: str
    target: str
    type: str


class ComparisonShortestPath(BaseModel):
    found: bool
    hops: int | None = None
    nodes: list[ComparisonPathNode] = Field(default_factory=list)
    links: list[ComparisonPathLink] = Field(default_factory=list)
    kind: Literal[
        "shared_exhibition",
        "shared_group",
        "shared_location",
        "general",
        "none",
    ]
    is_fallback: bool


class ArtistComparisonResponse(BaseModel):
    artist_a: ComparisonArtist
    artist_b: ComparisonArtist
    embedding_similarity: float
    same_cluster: bool
    common: ComparisonCommon
    unique: ComparisonUnique
    differences: list[ComparisonDifference] = Field(default_factory=list)
    co_exhibited: CoExhibitedSummary
    shortest_path: ComparisonShortestPath
    note: str
